### DISCUSSION TITLE
images referenced in CSS?

### DISCUSSION THREAD

**Discussion by jrochkind** [ElMassimo/vite_ruby#321] [Q&A] [answered]
I am having trouble figuring out how to succeed referencing local images in css, in a `url()`. 

I am using vite-ruby/vite-rails with rails. 

Can anyone provide a working example?  Where is a good place to put local images? 

Do I need to somehow register them (or the directory they are in), so that vite will know to build and serve them?  Or, wait... am I supposed to put them all in `./app/frontend/entrypoints`?  That seems a lot to me! 

If this is documented somewhere I'm missing, feel free to link me!

**Comment by ElMassimo** [comment, marked as answer]
Whenever an asset is referenced using `url` in CSS or imported from JS, Vite will fingerprint it and include it in the build.

Nothing special is required for referencing images inside CSS files, but I highly recommend using [the default import aliases](https://vite-ruby.netlify.app/guide/development.html#import-aliases-%F0%9F%91%89) or [your own import aliases](https://vite-ruby.netlify.app/guide/migration.html#assets-%F0%9F%8E%A8) to reference them:

```css
.background {
  background-image: url('@/images/bg.png');
}
```

The reason I recommend using aliases, and the `@` prefix in particular, is that it works across pre-processors, and Vite will always detect and transform it correctly.

---

On the other hand, if you want to [reference assets using tag helpers](https://vite-ruby.netlify.app/guide/rails.html#tag-helpers-%F0%9F%8F%B7), Vite needs to know that it should process them somehow.

A good location to place images is `app/frontend/images`, which is included by default in [`additionalEntrypoints`](https://vite-ruby.netlify.app/config/index.html#additionalentrypoints):

```erb
<%= vite_image_tag 'images/logo.jpg' %>
```

See [this section](https://vite-ruby.netlify.app/guide/advanced.html#additional-entrypoints) for an in-depth explanation.

**Reply by jrochkind** [reply]
Thank you! And speedier than expected response, as usual!

So, I have gotten it to work by putting the images in `./app/frontend/entrypoints/images`, but I agree that's kind of messy and `./app/frontend/images` would be better. Let me see if I can figure out what was going wrong there from your instructions. 

Aha! 

If it is in `./app/frontend/entrypoints/images/x.png`, then it can be referenced as simply `url(images/x.png)`, without the `@`. But putting it in `./app/frontend/images/x.png` instead, it needs the `@`, `url(@/images/x.png)`. 

Does that make any sense? That's fine now that I see it. I guess I should really get in the habit of using `@` everywhere, I guess it doesn't hurt even in places it isn't required.
